frontend/simple/js/state.js

```
@@ -118,7 +171,7 @@ const actions = {
 
     const type = entry.type
     entry = Events[type].fromObject(entry, hash)
-    if (entry instanceof Events.HashableContract) {
+    if (entry instanceof Events.HashableContract && !state.contracts[contractId]) {
```

What's this for?

---

If you perform this action on a contract that you already have subscribed to you'll get duplicate events. This is relevant when streaming a contracts events to synchronize.

---

Ah, OK, I think this is related to [my comment above](https://github.com/okTurtles/group-income-simple/pull/225/files#r111299591) about skipping the first one.

I think that is the proper way to do it. I.e. update the code up there, and then this check won't be necessary. We should never receive the same contract or action twice.

I was thinking at first that if we wanted to check for this properly, we might do so on line 167, e.g. by calling `db.getLogEntry(contractId, hash)` and making sure it returns `null`. However, that isn't necessary because `db.addLogEntry` already performs this check for us.

However, note that the code doesn't check the result of `db.addLogEntry`. *That* is a bug. We should either throw an exception, or perhaps better, just return from the function early. Note that would mean that any listeners for events at the end of the `handleEvent` wouldn't get called... which is probably the correct behavior anyway, as we have no idea what their effect might be, and we don't want to call something that would break things if it's called twice.

---

So there is another case that also makes this needed and not skipping the first. If we skip the first as in most cases it would leave out the case of when we are joining a group for the first time (in which we need all events inclusive). I can make a special case for this. I think we should change the `since`  to `from` this way the name will suggest inclusion.